cmd/zoekt-sourcegraph-indexserver/main.go

```diff
@@ -1103,7 +1139,44 @@ func newServer(conf rootConfig) (*Server, error) {
 }
 
 func main() {
-	if err := rootCmd().ParseAndRun(context.Background(), os.Args[1:]); err != nil {
+	cmd := rootCmd()
+	if err := cmd.Parse(os.Args[1:]); err != nil {
+		log.Fatal(err)
+	}
+
+	debugFlagOverride := false
+
+	flag := cmd.FlagSet.Lookup("debug")
+	if debugFlag, err := strconv.ParseBool(flag.Value.String()); err == nil {
+		// Debug flag overrides a non-debug logging level
+		debugFlagOverride = debugFlag && !srcLogLevelIsDebug()
+	}
+
+	name, _ := findName()
+	originalLogLevel := os.Getenv(sglog.EnvLogLevel)
+
+	// if debug flag overrides the logging level set by environment var EnvLogLevel
+	// then update it before we call logger's Init()
+	if debugFlagOverride {
+		os.Setenv(sglog.EnvLogLevel, "debug")
+	}
+
+	syncLogs := sglog.Init(sglog.Resource{
+		Name:       name,
+		Version:    zoekt.Version,
+		InstanceID: hostnameBestEffort(),
+	})
```

Comment:
is all the complexity due to us needing to support `-debug` flag? Maybe @bobheadxi has advice to make this not so clunky.

Alternatively we can deprecate the `-debug` flag and only support the envvar. In practice we only ever use the envvar.

Reply:
We currently don't have a way to set the base level besides the envvar, partly because we don't want multiple ways to configure the level (yet - there was some discussion about this in https://github.com/sourcegraph/sourcegraph/issues/33475#issuecomment-1103293853)

With your other comment in https://github.com/sourcegraph/zoekt/pull/374/files/538f85a77c8b81fe2e4fdea686571a9a4007589b#r910405965 that would simplify this significantly - I agree that resetting the env var is not needed, since it's only ever used on initialization

Reply:
@keegancsmith That's correct, the `-debug` flag and the resultant `debug` logger is why we have some extra complexity. The `debug` logger atm is a separate [instance](https://sourcegraph.com/github.com/sourcegraph/zoekt/-/blob/cmd/zoekt-sourcegraph-indexserver/main.go?L188) of the log libraries Logger. In utilizing sourcegraph/log  we're sharing the logging level ceiling across all loggers. 

Is there any concern if we stop a process with debug enabled via the program argument, and then start the process without the debug flag? In that case the env var would remain set to `debug`. I acknowledge that this might not be encountered for most folks but still wanted to get your thoughts.

I'll go ahead and update the PR to leave it set to debug if we overwrite the env var.

I'm also okay with removing the debug flag. My understanding after speaking with @stefanhengl a while back is that this is mainly used for our internal development purposes. That is, we don't use this for debugging a customer system or something like that.

Reply:
> The debug logger atm is a separate [instance](https://sourcegraph.com/github.com/sourcegraph/zoekt/-/blob/cmd/zoekt-sourcegraph-indexserver/main.go?L188) of the log libraries Logger. In utilizing sourcegraph/log we're sharing the logging level ceiling across all loggers.

You can use `(Logger).IncreaseLevel` or `(Logger).NoOp()`. The latter is safe to call pre-initialization so you can get the exact same behaviour with:

```
var debug = log.NoOp()
```

we also have an adapter for the std log package with `stdlog`

Reply:
IncreaseLevel was not able to help out in our case since we're potentially downgrading level with the debug flag. We could set the env var at the same place where we assign the debug logger to not discard [here](https://sourcegraph.com/github.com/sourcegraph/zoekt@6ba33af2cc5f7deec8f6caec00f2634921defd2e/-/blob/cmd/zoekt-sourcegraph-indexserver/main.go?L1099), but then we'd have to do sourcegraph/log `Init()` after this env var modification and I believe we should `Init()` in `main()` before we initialize our `Server` struct with `NewServer()`, not after. This challenge is why I ended up splitting `ParseAndRun()`

An alternative which uses IncreaseLevel would be to always set the env var to debug log level, call logger `Init()`, revert the env var, and then our main Scoped logger will call IncreaseLevel() as necessary whenever we initialize it outside of `main()` (e.g. in `NewServer()`). We'd be able to keep `ParseAndRun()` as is which simplifies some of the added complexity in my PR (i.e. we can `Init(`) the sourcegraph/log without looking at the program args), but we'd still have the extra logic of setting the environment variable. This also requires that all calls to create a Scoped Logger must call IncreaseLevel to match the actual src log level, and not the debug level that we temporarily set the src log level env var to in main(). I recognize that we should not be creating many scoped loggers, and instead pass around one, but in case we end up with some code where we cannot [propagate](https://docs.sourcegraph.com/dev/how-to/add_logging#scoped-loggers) a logger then we'll need to call scoped Logger multiple times.

Removing debug flag would probably be the best approach so I'll discuss with folks. We could implement that removal in this PR but that'd expand the gh issue scope so I'll need to discuss with others. 

I'll remove the package logger and will move the instance of scoped logger to the Server struct as well which will make it easy to pass around / propagate (similar to `[this](https://github.com/sourcegraph/zoekt/pull/411/files#diff-35d69225aaa31faa8c79a7b91fcaf29872ec68408e9b2fd48ed77151c0c4e71eR388)`).

Reply:
@keegancsmith I thought setenv might be overwriting config files, etc. and persisting the overwritten env var key value pair, my mistake. You can ignore my earlier concern about that. Since the env var is not used anywhere else as you both pointed out, I'll proceed with simplifying the logic to just keep it as `debug` after overwrite

Reply:
This thread is resolved: the debug flag is removed, so the additional complexity in `main()` is also removed. We still have the debug logger from the log package since we are not replacing all logging in this PR; we decide if this logger is no op based on the environment variable only, now that the flag is removed.